nimpy.nim

```diff
@@ -685,27 +689,57 @@ proc nimTupleToPy[T](o: T): PPyObject =
         discard pyLib.PyTuple_SetItem(result, i, nimValueToPy(f))
         inc i
 
-proc parseArg[T](argTuple: PPyObject, argIdx: int, result: var T) =
-    pyObjToNim(pyLib.PyTuple_GetItem(argTuple, argIdx), result)
-
-proc parseArg[T](argTuple: PPyObject, argIdx: int, default: T, result: var T) =
-    if argIdx < pyLib.PyTuple_Size(argTuple):
-        pyObjToNim(pyLib.PyTuple_GetItem(argTuple, argIdx), result)
-    else:
-        result = default
-
-proc verifyArgs(argTuple: PPyObject, argsLen, argsLenReq: int, funcName: string): bool =
+proc setPyArg[T](argTuple: PPyObject, argIdx: int, argVal: var T): bool =
+    if not argTuple.isNil:
+        if argIdx < pyLib.PyTuple_Size(argTuple):
+            pyObjToNim(pyLib.PyTuple_GetItem(argTuple, argIdx), argVal)
+            result = true
+
+proc setPyKwarg[T](kwargsDict: PPyObject, argName: cstring, argVal: var T): bool =
+    if not kwargsDict.isNil:
+        let item = pyLib.PyDict_GetItemString(kwargsDict, argName)
+        if not item.isNil:
+            pyObjToNim(item, argVal)
+            result = true
+
+proc parseArg[T](argTuple, kwargsDict: PPyObject, argIdx: int, argName: cstring, result: var T) =
+    let isArgSet = setPyArg(argTuple, argIdx, result)
+    if not isArgSet:
+        discard setPyKwarg(kwargsDict, argName, result)
+
+proc verifyArgs(argTuple, kwargsDict: PPyObject, argsLen, argsLenReq: int, argNames: openarray[cstring], funcName: string): bool =
     # WARNING! Do not let GC happen in this proc!
-    let sz = pyLib.PyTuple_Size(argTuple)
+    let
+        arg_sz = if argTuple.isNil: 0 else: pyLib.PyTuple_Size(argTuple)
+        kwarg_sz = if kwargsDict.isNil: 0 else: pyLib.PyDict_Size(kwargsDict)
+        sz = arg_sz + kwarg_sz
+        req_idx = max(0, argsLenReq - arg_sz - 1)
+
     result = if argsLen > argsLenReq:
         # We have some optional arguments, argsLen is the upper limit
         sz >= argsLenReq and sz <= argsLen
     else:
         sz == argsLen
+
     if not result:
         GC_disable()
         pyLib.PyErr_SetString(pyLib.PyExc_TypeError, funcName & "() takes exactly " & $argsLen & " arguments (" & $sz & " given)")
         GC_enable()
+        result = false
+
+    if not kwargsDict.isNil:
+        for k in keys(kwargsDict):
+            if $k notin argNames:
```

Comment:
This needs some more work.
1. `$` allocates and can trigger GC.
2. `argNames[req_idx..^1]` allocates and can trigger GC.